definitions/scenarios/self_upgrade.rb

```diff
@@ -0,0 +1,90 @@
+module ForemanMaintain::Scenarios
+  class SelfUpgradeBase < ForemanMaintain::Scenario
+    def enabled_system_repos_id
+      feature(:system_repos).enabled_repos_ids
+    end
+
+    def enable_repos(repo_ids = stored_enabled_repos_ids)
+      add_step(Procedures::Repositories::Enable.new(repos: repo_ids))
+    end
+
+    def disable_repos(repo_ids = stored_enabled_repos_ids)
+      add_step(Procedures::Repositories::Disable.new(repos: repo_ids))
+    end
+
+    def target_version
+      @target_version ||= context.get(:target_version)
+    end
+
+    def current_version
+      feature(:instance).downstream.current_minor_version
+    end
+
+    def maintenance_repo(version)
+      @maintenance_repo ||= "rhel-#{el_major_version}-server-satellite-maintenance-#{version}-rpms"
+    end
+
+    def stored_enabled_repos_ids
+      unless defined?(@stored_enabled_repos_ids)
+        @stored_enabled_repos_ids = []
+        backup_dir = File.expand_path(ForemanMaintain.config.backup_dir)
+        File.open(File.join(backup_dir, 'enabled_repos.yml'), 'r') do |repos_file|
+          @stored_enabled_repos_ids = YAML.load(repos_file.read)
+        end
```

Comment:
This file doesn't have to exist. `BackupEnabledRepos` only writes it if it's non-empty. I think this would fail then. It also won't be in older backups. You probably need to deal with that to prevent exceptions. You could also use `File.read` instead of `File.open` + `.read`, but that's a small nit.
```suggestion
        path = File.expand_path('enabled_repos.yml', ForemanMaintain.config.backup_dir)
        @stored_enabled_repos_ids = File.file?(path) ? YAML.load(File.read(path)) : []
```